### PR TITLE
Oobranch b : add utilies

### PR DESCRIPTION
**Pull request by antalk2**
Adds some general and UNO utilities

### CODE REVIEW

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
+package org.jabref.model.openoffice.util;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/*
+ * error cannot be null
+ * result cannot be null
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/util/OOResult.java:9]: Internal variable, please comment at the variables

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+public class OOResult<R, E> {
+    private final Optional<R> result;
+    private final Optional<E> error;
+
+    /**
+     * Exactly one of the arguments should be Optional.empty()
+     *
+     * @param result
+     * @param error
+     */
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/util/OOResult.java:24]: Don't add empty parameter comments

```suggestion
     */
```

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    private OOResult(Optional<R> result, Optional<E> error) {
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/util/OOResult.java:25]: Please, never ever use `Optional` as method parameter.

The most prominent way to fix is to offer different methods.

You already have these methods, thus just remove this method.

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+     * @param error Null is not allowed.
+     */
+    public static <R, E> OOResult<R, E> error(E error) {
+        return new OOResult<>(Optional.empty(), Optional.of(error));
+    }
+
+    /*
+     * Test state
+     */
+
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/util/OOResult.java:47]: Please no empty line between a JavaDoc and a method

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    /**
+     * @param result Null is not allowed.
+     */
+    public static <R, E> OOResult<R, E> ok(R result) {
+        return new OOResult<>(Optional.of(result), Optional.empty());
+    }
+
+    /**
+     * @param error Null is not allowed.
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/util/OOResult.java:38]: You can remove this comment. In nearly 99% of the cases, `null` is not allowed. If you really mean `Null`, please reference that class by using JavaDoc's `{@link ...}.

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+     */
+    public static <R, E> OOResult<R, E> error(E error) {
+        return new OOResult<>(Optional.empty(), Optional.of(error));
+    }
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/util/OOResult.java:42]: When not using Optionals as method parameters, the following code should do it:

```suggestion
    private <R, E> OOResult<R, E> of(E error) {
        this.result = Optional.empty();
        this.error = Optional.of(error);
    }

    public static <R, E> OOResult<R, E> of(E error) {
        return new OOResult<>(error);
    }
```

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        return !isError();
+    }
+
+    /*
+     * getters
+     */
+
+    public R get() {
+        if (isError()) {
+            throw new RuntimeException("Cannot get from error");
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/util/OOResult.java:70]: Never ever throw a `RuntimeException`. Please use `IllegalStateException` (even though this is also not good, but I see that it is required).

However, this is still not consistent to Optionals itself. You should throw a NullPointerException inssted.

Thus:

```java
    return result.orElseThrow(new NullPointerException("Cannot get from error"));
```

File: src/main/java/org/jabref/model/openoffice/util/OOTuple3.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package org.jabref.model.openoffice.util;
+
+public class OOTuple3<A, B, C> {
```
Comment (koppor) [inline src/main/java/org/jabref/model/openoffice/util/OOTuple3.java:3]: Can you please comment on the usage of the class? (JavaDoc)

**Review by koppor — COMMENTED** [review]
I quickly jumped to a random file and commented on that.

I think, these comments are general ones and could appear everywhere. Maybe you can check that.

**Comment by koppor** [comment]
@antalk2 Thank you for keeping working on it. For most of the developers it is still hard to follow. I am not sure whether you have seen our pull request template: https://github.com/JabRef/jabref/blob/main/.github/PULL_REQUEST_TEMPLATE.md

```markdown
<!-- 
Describe the changes you have made here: what, why, ... 
Link issues that are fixed, e.g. "Fixes #333".
If you fixed a koppor issue, link it, e.g. "Fixes https://github.com/koppor/jabref/issues/47".
The title of the PR must not reference an issue, because GitHub does not support autolinking there.
-->


<!-- 
- Go through the list below. If a task has been completed, mark it done by using `[x]`.
- Please don't remove any items, just leave them unchecked if they are not applicable.
-->

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.
```

I marked the part which is important to me in yellow:

grafik

I know, there were some discussions ago. However, I do not remember fully. Other developers (students, ...) browsing the JabRef repository may not know everything which happens in the JabRef world. Therefore, it would be good to have a long description in the first pull request - and then link from the other pull requests to that description. That would help others to understand the thing which is going on.

**Comment by koppor** [comment]
I propose following workflow:

0. `i:=b`
1. Create PR i **with a single commit**
2. Get feedback on PR i.
3. Incorporate feedback on PR i
4. PR i is merged into `main`
5. Update PR i+i
   - `git merge upstream/main`. Possibly, conflicts need to be resolved
6. Create **a single commit**
   1. Open `gitk --all` in case something goes wrong
   2. `git checkout branch-i+1`
   1. `git merge upstream/main`
   2. `git reset upstream/main`
   3. `git commit -m"PR i+1"` // maybe use a speaking commit message
   4. `git push -f`
7. `i++`; goto step 1

Since PRs i, i+1, ... already have been created, step 1 needs to be omited.

**Comment by Siedlerchr** [comment]
Your whole formatting looks a bit odd, do you have a maxline length configured?

**Comment by koppor** [comment]
This is the follow-up of https://github.com/JabRef/jabref/pull/7469.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> This is the follow-up of #7469.

This is a follow-up to https://github.com/koppor/jabref/pull/496, in particular https://github.com/koppor/jabref/pull/496#issuecomment-835377989.

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 1 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:9]: I thought of these more like stating class requirements.
`error` and `result` here can also be understood as constructor parameters.

Replaced with

```
An instance of this class represents either the result
of a computation, or an error value. Neither of these is allowed
to be null.
```

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 2 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:24]: ok, removed. But I am editing my working branch, and will copy then push.
To avoid confusion I did not click on the "commit suggestion button"

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 5 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:38]: removed

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 6 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:42]: I assume the `of` in the green line 40 was meant to be `OOResult` (constructor)

Following that line of thought I got to

```
    private OOResult(Optional<R> result) {
        this.result = result;
        this.error = Optional.empty();
    }

    private OOResult(Optional<E> error) {
        this.result = Optional.empty();
        this.error = error;
    }
```

I think this would be ambiguous if the types E and R are related.
What if both are Integer? Did I misunderstand the suggestion?

Green line 45: `of` ? I assume this comes from copying, not intended.

Keeping as is, for now. Need clarification.

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 3 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:25]: Stuck at ambiguity problem, see below. https://github.com/JabRef/jabref/pull/7787#discussion_r645172827
Could probably use constructor with no arguments. Would that be better?

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 4 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:47]: My intent was to express, that it does not belong to the next method,
but the next group (of four). A quick search suggests
[here](https://stackoverflow.com/questions/261876/grouping-javadoc-methods)
that javadoc does not out of the box support documenting a group together.

Am I mistaken? Should I remove the comment?

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 7 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:70]: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html#get()

Optional.get() throws NoSuchElementException

changed to `throw new NoSuchElementException("Cannot get from error");` to keep in line with Optional

File: src/main/java/org/jabref/model/openoffice/util/OOTuple3.java (hunk 8 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOTuple3.java:3]: I am not sure what do you mean here.
Inserted this:

```
/**
 * This class allows three objects to be packed together, and later
 * accessed as fields `a`, `b` and `c`.
 *
 * Can be used to avoid creating a new class for just this purpose.
 *
 * Can be useful if you do not have `Trifunction` at hand but need to
 * pass three objects at a time to a lambda function.
 *
 */
```

It is used in [RangeOverlapBetween.java](https://github.com/antalk2/jabref/blob/122d5133fa6c7b44245c5ba5600d398775718664/src/main/java/org/jabref/model/openoffice/rangesort/RangeOverlapBetween.java#L42)
to pack three values together without creating a class for that specific occasion.

Historically I created this class, then moved to ...Test when that was the only location
it was used, then eliminated its use, removed, then I needed it again. So here it is.
Not an absolute necessity, but I repeatedly needed it.

**Comment by antalk2** [comment]
I apologise, at first I did not read this part:

> I know, there were some discussions ago.
> However, I do not remember fully. 
> Other developers (students, ...) browsing the JabRef repository may not know everything which happens in the JabRef world.
> Therefore, it would be good to have a long description in the first pull request - and 
> then link from the other pull requests to that description.
> That would help others to understand the thing which is going on.

> a long description in the first pull request

Are you suggesting to (1) add a file to oobranch-A,  (2) under jabref/docs  ?

**Comment by antalk2** [comment]
> I propose following workflow:
> 
>     1. `i:=b`
> 
>     2. Create PR i **with a single commit**
> 
>     3. Get feedback on PR i.
> 
>     4. Incorporate feedback on PR i
> 
>     5. PR i is merged into `main`
> 
>     6. Update PR i+i
>        
>        * `git merge upstream/main`. Possibly, conflicts need to be resolved
> 
>     7. Create **a single commit**
>        
>        1. Open `gitk --all` in case something goes wrong
>        2. `git checkout branch-i+1`
>        3. `git merge upstream/main`
>        4. `git reset upstream/main`
>        5. `git commit -m"PR i+1"` // maybe use a speaking commit message
>        6. `git push -f`
> 
>     8. `i++`; goto step 1
> 
> 
> Since PRs i, i+1, ... already have been created, step 1 needs to be omited.

I am not sure why did you write this. 

- What is the problem you are trying to solve?
- The expression **a single commit** is emphasized.
- After updating and pushing PR i, I could not push PR i+1 without merging origin/PR i+1. I am not sure why. May be related 
to sending the PR.

**Comment by antalk2** [comment]
> Your whole formatting looks a bit odd, do you have a maxline length configured?

- Are you referring to some IntelliJ IDEA option?
- My editor is not configured to automatically break lines during typing
- Sometimes I request an auto-fill (I think only, or at least usually in comments) that does observe
some line length limit.

**Comment by koppor** [comment]
> * What is the problem you are trying to solve?

Making it both easy for you to update PRs i+j and easy for us to review the PRs.

>     * The expression **a single commit** is emphasized.

This is somehow low prio. We are somehow overwhelmed by the number of commits:

grafik

When looking into a commit, it seems, the commit changes code introduced in a previous commit. Thus, I personally wonder why the new commit is there:

grafik

If that is too much work for you, just keep the commits as is. We just work with "Files changed" view and scroll through there:

grafik

>     * After updating and pushing PR i, I could not push PR i+1 without merging origin/PR i+1. I am not sure why. May be related to sending the PR.

Not sure, why this happens. PR i should be unrelated to PR i+1, because the reference to the commit of PR i+1 does not change at all when updating PR i.

**Comment by koppor** [comment]
> > Your whole formatting looks a bit odd, do you have a maxline length configured?
> 
>     * Are you referring to some IntelliJ IDEA option?

Yes. Quoting from https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace:

grafik

(I posted the screenshot here as the image was 404 10 minutes ago, but the text was there)

>     * My editor is not configured to automatically break lines during typing

Example:

grafik

>     * Sometimes I request an auto-fill (I think only, or at least usually in comments) that does observe
>       some line length limit.

May, this is what we observe. The thing is that the comment appears much shorter than the code below. I think, we rely on rendering by IntelliJ (<kbd>Ctrl</kbd>+<kbd>Q</kbd>)

**Comment by antalk2** [comment]
> This is somehow low prio. We are somehow overwhelmed by the number of commits:
> 
> grafik

- I was wondering what those red x-es mean. Probably that those do not belong to this PR. You are right, I separated "general" and UNO because I assumed
such a grouping is helpful for you.
- Xlint: the compiler suggested I should check the code and at this point I figured out how to turn it on. 

> When looking into a commit, it seems, the commit changes code introduced in a previous commit. Thus, I personally wonder why the new commit is there:

> grafik

As far as I can remember, the generic programmic in java had some earlier
version, where a template without <> had a meaning, but turned out
not to be type-safe. Still accepts for backward compatibiity, but with
Xlint:unchecked suggests to change it.

> If that is too much work for you, just keep the commits as is.
Before the first push I can collapse, but each push is going to be
a distinct commit.

- I was wondering about when to push again
to oobranch-b. Until I do that, you do not see if it is OK now or not.

- related: when should I merge oobranch-b with upstream/main? If I understood
correctly, oobranch-a is already in the latter.

File: src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+     * @return The XTextRange for the bookmark, or Optional.empty().
+     */
+    public static Optional<XTextRange> getAnchor(XTextDocument doc, String name)
+        throws
+        WrappedTargetException,
+        NoDocumentException {
+
+        XNameAccess nameAccess = getNameAccess(doc);
+        return (UnoNameAccess.getTextContentByName(nameAccess, name)
+                .map(e -> e.getAnchor()));
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java:48]: Prefer static calls in things like map(Class::method) if you don't need to pass any parameters

File: src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+     *
+     *         result.getName() should be checked by the
+     *         caller, because its name may differ from the one
+     *         requested.
+     */
+    public static XNamed create(XTextDocument doc, String name, XTextRange range, boolean absorb)
+        throws
+        IllegalArgumentException,
+        CreationException {
+        return UnoNamed.insertNamedTextContent(doc,
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java:72]: It would be really nice if you could reformat your code without this line wrappings.

File: src/main/java/org/jabref/model/openoffice/uno/UnoCast.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+     * unoQI : short for UnoRuntime.queryInterface
+     *
+     * @return A reference to the requested UNO interface type if available,
+     *         otherwise null
+     */
+    public static <T> T unoQI(Class<T> zInterface, Object object) {
+        return UnoRuntime.queryInterface(zInterface, object);
+    }
+
+    public static <T> Optional<T> optUnoQI(Class<T> zInterface, Object object) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/openoffice/uno/UnoCast.java:21]: I don't understand why you need both methods

File: src/main/java/org/jabref/model/openoffice/uno/UnoCrossRef.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        // based on: https://wiki.openoffice.org/wiki/Documentation/DevGuide/Text/Reference_Marks
+        XMultiServiceFactory msf = UnoCast.unoQI(XMultiServiceFactory.class, doc);
+        // Create a 'GetReference' text field to refer to the reference mark we just inserted,
+        // and get it's XPropertySet interface
+        XPropertySet xFieldProps;
+        try {
+            String name = "com.sun.star.text.textfield.GetReference";
+            xFieldProps = (XPropertySet) UnoCast.unoQI(XPropertySet.class,
+                                                       msf.createInstance(name));
+        } catch (Exception e) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/openoffice/uno/UnoCrossRef.java:53]: Is it possible to catch a more specific Exception? You

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 3 above)
Comment (koppor) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:25]: With following comment

> When not using Optionals as method parameters, the following code should do it:

I meant:

Please do not use `Optionals` as parameter. Please use my suggested code.

Comment (koppor) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:25]: Reading further: This seems to be impossible. Then, just leave as is.

Comment (koppor) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:25]: Or use `ofError` and `ofResult`.

File: src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java (hunk 9 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java:48]: ok.
At one location I could not do this:

`.map(e -> OOText.asString(e))`

Originally it was

`.map(e -> OOText.toString(e))`, but the compiler complained about `OOText::toString` being ambiguous
with `Object.toString`. Renamed to `asString` now it is ambiguous with the non-static `asString`. At least now
both are visible in OOText.java

Could rename  the non-static `asString` to `getString` if necessary. Or override `Object.toString`, 
but I do not feel that the string in the box '"textually represents"' the boxed version.

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 4 above)
Comment (koppor) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:47]: Ah, yeah. It is not common to add comments on groups of methods. In the concrete case, I would remove the comment as "Test state" is not that clear (at least to me). Comment using the other extreme would be: Write a whole paragraph if the comment is really important... - In the context of JabRef, we are writing very little comments as we think that comments might be outdated soon and maintainers do not want update comments.

In the book "Java-by-Comparison", there is a whole example on comments: http://media.pragprog.com/titles/javacomp/examples.pdf.

**Comment by koppor** [comment]
>     * I was wondering about when to push again
>       to oobranch-b. 

After each fix you do.

> Until I do that, you do not see if it is OK now or not.

Yes. From that, you can merge into your oobranch-c (or "just" into your local branch). We do not want to increase the effort that much.
 
>     * related: when should I merge oobranch-b with upstream/main? 

If some non-related tests are red, this can help. Otherwise, not necessary.

> If I understood correctly, oobranch-a is already in the latter.

Yeah! We want to move things forward :).

File: src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java (hunk 9 above)
Comment (koppor) [reply src/main/java/org/jabref/model/openoffice/uno/UnoBookmark.java:48]: In that case, I would leave the `toString(e)`, because this is a standard way of Java to convert to a string. This overrules the "static calls" guide :)

File: src/main/java/org/jabref/model/openoffice/uno/UnoCast.java (hunk 11 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/uno/UnoCast.java:21]: unoQI was born to let me see the varying part.

```
XMultiServiceFactory msf = UnoRuntime.queryInterface(XMultiServiceFactory.class, doc);
XMultiServiceFactory msf = unoQI(XMultiServiceFactory.class, doc);
```

When breaking up the UNO-related collection, it became:

```
XMultiServiceFactory msf = UnoCast.unoQI(XMultiServiceFactory.class, doc);
```

At the time I either did not know about static imports,
or was not sure if it is accepted in JabRef code.

`import static` appears in the codebase, mostly in tests, occasionally elsewhere too.
But it does not tell me if it is in the "we are sorry, did not have time to remove"
or in the "yes, we love that category". 

Question: is `import static` allowed in JabRef? This would allow me to drop
the `UnoCast.` prefixes.

`optUnoQI` was born when I adjusted to a request: `do not return null`

It allows for example
```
return UnoCast.optUnoQI(XPropertySet.class, propertyContainer);
```
instead of
```
return Optional.ofNullable(UnoCast.unoQI(XPropertySet.class, propertyContainer));
```
or
```
return Optional.ofNullable(UnoRuntime.queryInterface(XPropertySet.class, propertyContainer));
```





It is not used in as many locations as I expected: in some cases
the cast is known (or usually expected) to be successful (the input is documented
to support the interface), in other cases internally I kept the testing for null
instead of converting to Optional then asking for isPresent

- UnoCast.unoQI : 40 matches
- UnoCast.optUnoQI : 6 matches

Comment (Siedlerchr) [reply src/main/java/org/jabref/model/openoffice/uno/UnoCast.java:21]: In tests import Static is okay, but otherwise it shouldn't be really used.
If it's not too much effort, I would stick with the method that returns Optional so optUnoQI

File: src/main/java/org/jabref/model/openoffice/uno/UnoCrossRef.java (hunk 12 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/uno/UnoCrossRef.java:53]: https://www.openoffice.org/api/docs/common/ref/com/sun/star/lang/XMultiServiceFactory.html#createInstance

- `raises(  ::com::sun::star::uno::Exception );`

https://www.openoffice.org/api/docs/common/ref/com/sun/star/uno/Exception.html

- All exceptions defined in UNO idl should derive from this exception.

https://wiki.openoffice.org/wiki/Documentation/DevGuide/ProUNO/Java/Mapping_of_Exception_Types

```
package com.sun.star.uno;
   public class Exception extends java.lang.Exception [...]
```

yes. `com.sun.star.uno.Exception` is not the same as `java,lang.Exception`
Modified. to `catch (con.sun.star.uno.Exception e)`

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 4 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:47]: Dropped group comments

File: src/main/java/org/jabref/model/openoffice/util/OOResult.java (hunk 3 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/util/OOResult.java:25]:
```
return OOResult.ok(cursor);
return OOResult.error(message);
```
vs
```
return OOResult.ofResult(cursor);
return OOResult.ofError(message);
```
The first pair seems more suggestive of the important part: success or failure.
The part that we are constructing something (`of`) is just background, not worth spelling out.

File: src/main/java/org/jabref/model/openoffice/uno/UnoCast.java (hunk 11 above)
Comment (antalk2) [reply src/main/java/org/jabref/model/openoffice/uno/UnoCast.java:21]: Dropped unoQI, renamed optUnoQI to cast

**Comment by antalk2** [comment]
Was: more detailed description

I wrote something [here](https://github.com/antalk2/jabref/tree/improve-reversibility-rebased-03/docs/openoffice)

- Does it cover what you expected?
- Shall I copy it to oobranch-B ?

**Comment by Siedlerchr** [comment]
@antalk2  Thanks so much for the detailed write up! That is very helpful! From my point of view this PR here is ready codewisem so we can continue with C. I am just thinking about how to proceed so the quesiton is:

Can this be merged into main without breaking things or should this be integrated first into branch C?

**Comment by antalk2** [comment]
> @antalk2 Thanks so much for the detailed write up! That is very helpful! From my point of view this PR here is ready codewisem so we can continue with C. I am just thinking about how to proceed so the quesiton is:
> 
> Can this be merged into main without breaking things or should this be integrated first into branch C?

Apart from moving
 src/main/java/org/jabref/gui/openoffice/CreationException.java
 src/main/java/org/jabref/gui/openoffice/NoDocumentException.java 
to
 src/main/java/org/jabref/model/openoffice/uno/CreationException.java
 src/main/java/org/jabref/model/openoffice/uno/NoDocumentException.java

and adding
```
public NoDocumentException() {
            super("Not connected to a document");
}
```
and changing the import statements accordingly in 
src/main/java/org/jabref/gui/openoffice/OOBibBase.java and
src/main/java/org/jabref/gui/openoffice/OpenOfficePanel.java

there are only additions of new files. So it is unlikely to break anything.

Breaking things can come in "integration", where we swap (calls to old code)
to (calls to new code).

Shall I leave https://github.com/antalk2/jabref/tree/improve-reversibility-rebased-03/docs/openoffice where it is,
or should I copy it into oobranch-B (or elsewhere)?

**Comment by Siedlerchr** [comment]
Okay thanks for the work! We merged is into main now, so we can continue reviewing the next branch then
